### PR TITLE
[GSProcessing] Add pre-computed categorical transformation loading

### PR DESCRIPTION
**Pull request by thvasilo**
*Issue #, if available:*

*Description of changes:*

* Follow-up to #857 
* Allow us to re-apply a previously saved categorical transformation to new data. See below for design details.

To be able to re-apply the categorical transformations that we create using the code in #857 , we first create a mapping from original string to one-hot representation, that we read from the saved JSON file, then use a UDF to use the mapping(s) on the column(s).

The `DistributedTransformation` class from which all transformation implementations inherit, gains a new function, `apply_precomputed_transformation`. When a pre-computed transformation JSON file exists in the input, and the feature is one of those listed in that file, we use this function to re-apply the existing transformation instead of creating a new one.

The default implementation for `apply_precomputed_transformation` is to log a warning and apply a new transformation. 

When we implement a pre-computed transform for a new transformation (e.g. numerical) we need to:
* Ensure the the transformation's `self.json_representation` is populated during the call to `apply()`. This ensures the transformation info will be saved in the output JSON.
* Override the `apply_precomputed_transformation` function (as we did for `DistCategoryTransformation` here), so that it uses the dict loaded from the JSON file to re-apply the transformation to the new data.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: graphstorm-processing/graphstorm_processing/distributed_executor.py
Comment (thvasilo) [inline graphstorm-processing/graphstorm_processing/distributed_executor.py:1]: Changes to this file are mostly cosmetic, e.g. we moved the initialization of the `DistHeterogeneousGraphLoader` object from the `run` function to the DistributedExecutor's __init__, ensuring a `loader` class variable exists the moment the `DistributedExecutor` object is created.

File: graphstorm-processing/graphstorm_processing/graph_loaders/dist_heterogeneous_loader.py [hunk 1]
```diff
@@ -1029,16 +1036,26 @@ def _process_node_features(
...
         node_type_feature_metadata = {}
         ntype_feat_sizes = {}  # type: Dict[str, int]
+
         for feat_conf in feature_configs:
-            json_representation = self.transformation_representations.get(feat_conf.feat_name, {})
+            json_representation = (
+                self.pre_computed_transformations.get("node_features", {})
+                .get(node_type, {})
+                .get(feat_conf.feat_name, {})
+            )
```
Comment (thvasilo) [inline graphstorm-processing/graphstorm_processing/graph_loaders/dist_heterogeneous_loader.py:1045]: This will get a value if there exists a pre-computed representation for this feature name and node type, None otherwise. We do the same for the edges.

File: graphstorm-processing/pyproject.toml [hunk 2]
```diff
@@ -40,6 +40,7 @@ black = "~24.2.0"
 pre-commit = "^3.3.3"
 types-mock = "^5.1.0.1"
 pylint = "~2.17.5"
+diff-cover = "^9.0.0"
```
Comment (thvasilo) [inline graphstorm-processing/pyproject.toml:43]: I use this to determine how test coverage changes between commits.

File: graphstorm-processing/graphstorm_processing/data_transformations/dist_transformations/base_dist_transformation.py [hunk 3]
```diff
@@ -52,6 +62,28 @@ def get_json_representation(self) -> dict:
         else:
             return {}
 
+    def apply_precomputed_transformation(self, input_df: DataFrame) -> DataFrame:
```
Comment (thvasilo) [inline graphstorm-processing/graphstorm_processing/data_transformations/dist_transformations/base_dist_transformation.py:65]: This function acts as a fail-safe, in case somehow we try to re-apply a transformation for a transformer that doesn't support re-applying.

This can only happen if a `precomputed_transformations.json` is modified by hand to include a value for a not-supported transformation.

The alternative is to raise a `NotImplementedError()` here, but I think this is safer.

Comment (jalencato) [reply graphstorm-processing/graphstorm_processing/data_transformations/dist_transformations/base_dist_transformation.py:65]: I did not fully understand the point here. I think we want to avoid to run the not-supported transformation when re-applying the transformation, right? But in this way we are still running the previous apply function on not-supported transformation, but `NotImplementedError()` will stop the whole process. Why will the method be safer?

File: graphstorm-processing/graphstorm_processing/data_transformations/dist_transformations/dist_category_transformation.py [hunk 4]
```diff
@@ -157,14 +165,90 @@ def apply(self, input_df: DataFrame) -> DataFrame:
...
+        per_col_label_to_one_hot_idx: dict[str, dict[str, int]] = self.json_representation[
+            "per_col_label_to_one_hot_idx"
+        ]
+        precomputed_cols: list[str] = self.json_representation["cols"]
+
+        # Assertions to ensure correctness of representation
+        assert set(precomputed_cols) == set(self.cols), (
+            f"Mismatched columns in precomputed transformation: "
+            f"pre-computed cols: {sorted(precomputed_cols)}, "
+            f"columns in current config: {sorted(self.cols)}"
```
Comment (jalencato) [inline graphstorm-processing/graphstorm_processing/data_transformations/dist_transformations/dist_category_transformation.py:189]: May we print out the diff here? I think the users will care more about the diff.

File: docs/source/gs-processing/usage/example.rst [hunk 5]
```diff
@@ -173,7 +173,7 @@ we can use the following command to run the processing job locally:
 
 .. code-block:: bash
 
-    gs-processing --config-filename gconstruct-config.json \
+    gs-processing --config-filename gsprocessing-config.json \
```
[the remark this answers is not in the document]
Comment (thvasilo) [reply docs/source/gs-processing/usage/example.rst:176]: Yes changed the config to the gsprocessing one as that one includes more features, and the category feature state, so the example makes more sense. I'll fix the follow-up example thanks.

File: graphstorm-processing/graphstorm_processing/data_transformations/dist_transformations/base_dist_transformation.py (hunk 3 above)
Comment (thvasilo) [reply graphstorm-processing/graphstorm_processing/data_transformations/dist_transformations/base_dist_transformation.py:65]: The idea here is to have is do apply a new transformation instead of errorring out, in the convoluted even that a non-supported transformation ends up in `precomputed_transformations.json`. The scenario is the following:


* A user runs gs-processing on some data, let's say they have a numerical feature, `age`, and a categorical feature, `state`.
* We save the transformation for `state` in `precomputed_transformations.json` under the output path.
* The user modifies `precomputed_transformations.json` to include the `age` feature, and copies the file to some new input data.
* When the user runs gs-processing on the new data, we'll read an entry for `age` in `precomputed_transformations.json`, and will call `DistNumericalTransformation.apply_precomputed_transformation()` to re-apply it.
* The default implementation will run here, printing a warning and creating a new transformation.

I realize though that the scenario that "The user modifies `precomputed_transformations.json` to include the `age` feature" is very unlikely. Maybe just having `NotImplementedError()` here is better? That way if a non-supported feature ends up in `precomputed_transformations.json` for any reason we just error out.

File: graphstorm-processing/graphstorm_processing/data_transformations/dist_transformations/dist_category_transformation.py (hunk 4 above)
Comment (thvasilo) [reply graphstorm-processing/graphstorm_processing/data_transformations/dist_transformations/dist_category_transformation.py:189]: Will do.

File: graphstorm-processing/graphstorm_processing/data_transformations/dist_transformations/base_dist_transformation.py (hunk 3 above)
Comment (jalencato) [reply graphstorm-processing/graphstorm_processing/data_transformations/dist_transformations/base_dist_transformation.py:65]: That makes sense.

File: graphstorm-processing/graphstorm_processing/data_transformations/dist_transformations/dist_category_transformation.py [hunk 6]
```diff
@@ -157,14 +165,91 @@ def apply(self, input_df: DataFrame) -> DataFrame:
...
             "per_col_label_to_one_hot_idx": per_col_label_to_one_hot_idx,
             "transformation_name": self.get_transformation_name(),
         }
 
         return dense_vector_features
 
+    def apply_precomputed_transformation(self, input_df: DataFrame) -> DataFrame:
+
+        # Get JSON representation of categorical transformation
+        labels_arrays: list[list[str]] = self.json_representation["string_indexer_labels_arrays"]
```
Comment (classicsong) [inline graphstorm-processing/graphstorm_processing/data_transformations/dist_transformations/dist_category_transformation.py:179]: Is this for task labels? and per_col_label_to_one_hot_idx is for features?
Can you add some comments to clarify it?

File: graphstorm-processing/graphstorm_processing/graph_loaders/dist_heterogeneous_loader.py [hunk 7]
```diff
@@ -49,16 +48,21 @@
     HUGGINGFACE_TOKENIZE,
     TRANSFORMATIONS_FILENAME,
 )
-from ..config.config_parser import EdgeConfig, NodeConfig, StructureConfig
-from ..config.label_config_base import LabelConfig
-from ..config.feature_config_base import FeatureConfig
-from ..data_transformations.dist_feature_transformer import DistFeatureTransformer
-from ..data_transformations.dist_label_loader import DistLabelLoader, SplitRates, CustomSplit
-from ..data_transformations import s3_utils, spark_utils
+from graphstorm_processing.config.config_parser import EdgeConfig, NodeConfig, StructureConfig
```
Comment (classicsong) [inline graphstorm-processing/graphstorm_processing/graph_loaders/dist_heterogeneous_loader.py:51]: Why change it to use graphstorm_processing.config.xxx ?

File: graphstorm-processing/graphstorm_processing/graph_loaders/dist_heterogeneous_loader.py [hunk 8]
```diff
@@ -1493,6 +1511,7 @@ def _process_edge_features(
             for each feature, which tells us the column size of the
             encoded features.
         """
+        # TODO: Add unit tests for edge feature processing
```
Comment (classicsong) [inline graphstorm-processing/graphstorm_processing/graph_loaders/dist_heterogeneous_loader.py:1514]: When will you add the tests?

File: graphstorm-processing/graphstorm_processing/graph_loaders/dist_heterogeneous_loader.py (hunk 7 above)
Comment (thvasilo) [reply graphstorm-processing/graphstorm_processing/graph_loaders/dist_heterogeneous_loader.py:51]: pylint was throwing errors about trying to import packages beyond the top level. This ensures the right thing happens regardless of how the package is imported

File: graphstorm-processing/graphstorm_processing/graph_loaders/dist_heterogeneous_loader.py (hunk 8 above)
Comment (thvasilo) [reply graphstorm-processing/graphstorm_processing/graph_loaders/dist_heterogeneous_loader.py:1514]: The file is at over 80% code coverage now, so I didn't consider that a priority. This is more of a future pointer about which parts to add tests for.

File: graphstorm-processing/graphstorm_processing/data_transformations/dist_transformations/dist_category_transformation.py (hunk 6 above)
Comment (thvasilo) [reply graphstorm-processing/graphstorm_processing/data_transformations/dist_transformations/dist_category_transformation.py:179]: No the naming comes from Spark and I agree is a bit confusing: https://spark.apache.org/docs/3.5.0/api/python/reference/api/pyspark.ml.feature.StringIndexerModel.html#pyspark.ml.feature.StringIndexerModel.labelsArray

Spark refers to unique values (strings) in a categorical column as "labels". 

I'm adding comments here to clarify.

File: docs/source/gs-processing/usage/example.rst [hunk 9]
```diff
@@ -211,26 +211,41 @@ and can be used downstream to create a partitioned graph.
...
 
     $ cd /tmp/gsprocessing-example
-    $ ls
-
-    edges/  launch_arguments.json  metadata.json  node_data/
-    raw_id_mappings/  perf_counters.json  updated_row_counts_metadata.json
+    $ ls -l
+
+    edges/
+    gconstruct-config_with_transformations.json
```
[the remark this answers is not in the document]
Comment (thvasilo) [reply docs/source/gs-processing/usage/example.rst]: Agree on the confusing part, but since we want to support cross-compatibility with GConstruct, it's the least invasive solution. 

Otherwise we'd have to re-parse the entire config file and do that for both versions of the configs. 

I'd prefer to keep it this way, one is intended for end users (`gconstruct-config_with_transformations.json`) the other just to be read by GSProcessing.

[the remark this answers is not in the document]
Comment (classicsong) [reply docs/source/gs-processing/usage/example.rst]: OK

**Comment by thvasilo** [comment]
Any concerns left for this PR?

**Comment by classicsong** [comment]
I am OK with the PR.
Please check with @jalencato.

**Review by classicsong — APPROVED** [review]

**Review by jalencato — APPROVED** [review]
LGTM
